design/repo_maintenance_job_config.md

```diff
@@ -0,0 +1,237 @@
+
+## Non Goals
+- There was an [issue](https://github.com/vmware-tanzu/velero/issues/7911) to require the whole Job's PodSpec should be configurable. That's not in the scope of this design.
+- Please notice this new configuration is dedicated for the repository maintenance. Repository itself configuration is not covered.
+
+
+## Compatibility
+v1.14 uses the `velero server` CLI's parameter to pass the repository maintenance job configuration.
+In v1.15, those parameters are removed, including `--maintenance-job-cpu-request`, `--maintenance-job-mem-request`, `--maintenance-job-cpu-limit` and `--maintenance-job-mem-limit`.
+Instead, the parameters are read from the ConfigMap specified by `velero server` CLI parameter `--repo-maintenance-job-config` introduced by this design.
+
+## Design
+This design introduces a new ConfigMap specified by `velero server` CLI parameter `--repo-maintenance-job-config` as the source of the repository maintenance job configuration. The specified ConfigMap is read from the namespace where Velero is installed.
+If the ConfigMap doesn't exist, the internal default values are used.
+
+Example of using the parameter `--repo-maintenance-job-config`:
+```
+velero server \
+    ...
+    --repo-maintenance-job-config repo-job-config
+    ...
+```
+
+**Notice**
+* Velero doesn't own this ConfigMap. If the user wants to customize the repository maintenance job, the user needs to create this ConfigMap.
+* Velero reads this ConfigMap content at starting a new repository maintenance job, so the ConfigMap change will not take affect until the next created job.
+
+### Structure
+The data structure for ```repo-maintenance-job-config``` is as below:
+```go
+type MaintenanceConfigMap map[string]Configs
+
+type Configs struct {
+    // LoadAffinity is the config for data path load affinity.
+    LoadAffinity []*LoadAffinity `json:"loadAffinity,omitempty"`    
+
+    // Resources is the config for the CPU and memory resources setting.
+    Resource     Resources       `json:"resources,omitempty"`
+}
+
+type LoadAffinity struct {
+    // NodeSelector specifies the label selector to match nodes
+    NodeSelector metav1.LabelSelector `json:"nodeSelector"`
+}
+
+type Resources struct {
+    // The repository maintenance job CPU request setting
+	CPURequest          string `json:"cpuRequest,omitempty"`
+
+    // The repository maintenance job memory request setting
+	MemRequest          string `json:"memRequest,omitempty"`
+
+    // The repository maintenance job CPU limit setting
+	CPULimit            string `json:"cpuLimit,omitempty"`
+
+    // The repository maintenance job memory limit setting
+	MemLimit            string `json:"memLimit,omitempty"`
+}
+```
+
+The ConfigMap content is a map.
+If there is a key value as `global` in the map, the key's value is applied to all BackupRepositories maintenance jobs that don't their own specific configuration in the ConfigMap.
+The other keys in the map should be the name of BackupRepository.
```

Comment:
BackupRepository  CR object is created by Velero and not by the users, So how would the users know the name of the BackupRepository object during the Velero install time ?

Reply:
Isn't there typically a pattern to how backuprepository name gets generated? Could doc that and user can follow.

Reply:
IMO, it's not possible to customize specific repository maintenance jobs at the Velero installation stage, because the repository name has a random suffix.

But there is a `global` section in this ConfigMap. The `global` section is applied to all BackupRepositories that don't have a specific section. If there is a need to fine-tuning a BackupRepository, the user can modifies the ConfigMap on the need.

Does that solution sound reasonable to you?
@shubham-pampattiwar

Reply:
@blackpiglet another possibility would be to use a key like "bslName-namespaceName" instead of backup repository name -- that could be predicted ahead of time, and since backup repos should be unique to BSL+namespace, it would be equivalent. In any case, I think the global section is most likely to handle the vast majority of use cases, and the per-repo config is more of an edge case.

Reply:
OK, the combination of the following triple keys should make sure only one instance.
```
type BackupRepositoryKey struct {
	VolumeNamespace string
	BackupLocation  string
	RepositoryType  string
}
```

Reply:
Modified.